File: gred/src/flag_manager.lua
FlagManager = class()

function FlagManager:Create()
  self = FlagManager:New()
  self.lst = {}
  self.current_active = nil
  return self
end

function FlagManager:update(active_flag)
  if active_flag ~= self.current_active then
    if self.current_active ~= nil then
      local flag = self:find(self.current_active)
      if flag ~= nil then flag:setActive(false) end
      self.current_active = nil
    end
    if active_flag ~= nil then
      local flag = self:find(active_flag)
      if flag ~= nil then
        flag:setActive(true)
        self.current_active = active_flag
      end
    end
  end
end

function FlagManager:drawFlagNumbers(font, cam)
  for _, flag in ipairs(self.lst) do
    WorldToScreen(cam, EntityX(flag.entity), EntityY(flag.entity), EntityZ(flag.entity))
    local width = TextWidth(font, flag.id)
    local height = TextHeight(font, flag.id)
    DrawRect(PointX(), PointY(), width, height, COLOR_BLACK)
    DrawText(font, flag.id, PointX(), PointY(), COLOR_YELLOW)
  end
end

function FlagManager:clear()
  for _, flag in ipairs(self.lst) do
    flag:destroy()
  end
  self.lst = {}
  self.current_active = nil
end

function FlagManager:place(flag_id, x, y, z)
  local index = self:findIndex(flag_id)
  if index == nil then
    self.lst[#self.lst + 1] = Flag:Create(flag_id, x, y, z)
    table.sort(self.lst, function(a, b) return a.id <= b.id end)
  else
    self.lst[index]:position(x, y, z)
  end
end

function FlagManager:remove(flag_id)
  local index = self:findIndex(flag_id)
  if index == nil then return nil end
  self.lst[index]:destroy()
  table.remove(self.lst, index)
end

function FlagManager:find(flag_id)
  local index = self:findIndex(flag_id)
  if index ~= nil then return self.lst[index] else return nil end
end

function FlagManager:findIndex(flag_id)
  for i, flag in ipairs(self.lst) do
    if flag.id == flag_id then return i end
  end
  return nil
end

function FlagManager:size()
  return #self.lst
end

function FlagManager:at(index)
  if index < 0 or index > self:size() then return nil end
  return self.lst[index]
end

File: src/flag_manager.cpp
#include "flag_manager.h"

FlagManager::FlagManager() {}

void FlagManager::update(std::optional<int> active_flag) {
    if (active_flag != current_active) {
        if (current_active.has_value()) {
            Flag* f = find(*current_active);
            if (f) f->set_active(false);
            current_active = std::nullopt;
        }
        if (active_flag.has_value()) {
            Flag* f = find(*active_flag);
            if (f) {
                f->set_active(true);
                current_active = active_flag;
            }
        }
    }
}

void FlagManager::draw_flag_numbers(Camera* cam) {
    for (Flag* flag : lst) {
        App::world_to_screen(cam, flag->x(), flag->y(), flag->z());
        std::string label = std::to_string(flag->id);
        int w = App::text_width(label);
        int h = App::text_height(label);
        App::draw_rect((int)App::point_x, (int)App::point_y, w, h, COLOR_BLACK);
        App::draw_text(label, (int)App::point_x, (int)App::point_y, COLOR_YELLOW);
    }
}

void FlagManager::clear() {
    for (Flag* f : lst) f->destroy();
    lst.clear();
    current_active = std::nullopt;
}

void FlagManager::place(int flag_id, float x, float y, float z) {
    auto idx = find_index(flag_id);
    if (!idx.has_value()) {
        lst.push_back(new Flag(flag_id, x, y, z));
        std::sort(lst.begin(), lst.end(), [](Flag* a, Flag* b){ return a->id < b->id; });
    } else {
        lst[*idx]->position(x, y, z);
    }
}

void FlagManager::remove(int flag_id) {
    auto idx = find_index(flag_id);
    if (!idx.has_value()) return;
    lst[*idx]->destroy();
    lst.erase(lst.begin() + *idx);
}

Flag* FlagManager::find(int flag_id) {
    auto idx = find_index(flag_id);
    return idx.has_value() ? lst[*idx] : nullptr;
}

std::optional<int> FlagManager::find_index(int flag_id) const {
    for (int i = 0; i < (int)lst.size(); ++i)
        if (lst[i]->id == flag_id) return i;
    return std::nullopt;
}

int FlagManager::size() const { return (int)lst.size(); }

Flag* FlagManager::at(int index) {
    if (index <= 0 || index > size()) return nullptr;
    return lst[index - 1];
}
